cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MigrationTableUtility.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.metrics.process;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.dataset2.lib.table.MetricsTable;
+import co.cask.cdap.data2.util.TableId;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.metrics.store.MetricDatasetFactory;
+import co.cask.cdap.proto.id.NamespaceId;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.TableNotFoundException;
+import org.apache.hadoop.hbase.client.HBaseAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Useful to get v2/v3 metrics tables with retry and check if metrics table exists or not
+ */
+public class MigrationTableUtility {
+  private static final Logger LOG = LoggerFactory.getLogger(MigrationTableUtility.class);
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final MetricDatasetFactory metricDatasetFactory;
+  private final int resolution;
+  private final HBaseTableUtil tableUtil;
+
+  MigrationTableUtility(CConfiguration cConfiguration, Configuration hConf,
+                        MetricDatasetFactory metricDatasetFactory, HBaseTableUtil tableUtil, int resolution) {
+    this.cConf = cConfiguration;
+    this.hConf = hConf;
+    this.metricDatasetFactory = metricDatasetFactory;
+    this.tableUtil = tableUtil;
+    this.resolution = resolution;
+  }
+
+  // we try to get v2 metrics table with retry, there is a chance that v2 table got deleted before we got it
+  // we return null if it was already deleted
+  @Nullable
+   MetricsTable getV2MetricsTable() {
+    MetricsTable v2table = null;
+    while (v2table == null) {
+      try {
+        v2table = metricDatasetFactory.getV2MetricsTable(resolution);
+      } catch (Exception e) {
+        if (e instanceof TableNotFoundException) {
+          // we checked table exists, however the table got deleted in-between before we got the table
+          return null;
+        }
+        LOG.warn("Cannot access v2 metricsTable, will retry in 1 sec.");
+        try {
+          TimeUnit.SECONDS.sleep(1);
```

Comment:
updated the logic to perform table get in scheduled run and not during initialization on this commit. https://github.com/caskdata/cdap/pull/9399/commits/0cf263b5c766af5e7e050680fb610160ea0b1a82